air/src/boundry.rs

```diff
@@ -0,0 +1,271 @@
+use ark_ff::PrimeField;
+use composition_polynomial::{air::Air, composition_polynomial::CompositionPolynomial};
+
+#[derive(Clone)]
+pub struct BoundaryAir<F: PrimeField> {
+    trace_length: usize,
+    n_columns: usize,
+    constraints: Vec<ConstraintData<F>>,
+    mask: Vec<(usize, usize)>,
+}
+
+#[derive(Clone)]
+pub struct ConstraintData<F> {
+    coeff_idx: usize,
+    column_index: usize,
+    point_x: F,
+    point_y: F,
+}
+
+impl<F: PrimeField> BoundaryAir<F> {
+    pub fn new(
+        trace_length: usize,
+        n_columns: usize,
+        boundary_conditions: &[(usize, F, F)],
+    ) -> Self {
+        let mut constraints = Vec::with_capacity(boundary_conditions.len());
+
+        for (coeff_idx, &(column_index, ref point_x, ref point_y)) in
+            boundary_conditions.iter().enumerate()
+        {
+            let x = *point_x;
+
+            let pos = constraints
+                .iter()
+                .position(|constraint: &ConstraintData<F>| constraint.point_x == x);
+
+            let constraint_data = ConstraintData {
+                coeff_idx,
+                column_index,
+                point_x: x,
+                point_y: *point_y,
+            };
+
+            if let Some(pos) = pos {
+                constraints.insert(pos, constraint_data);
+            } else {
+                constraints.push(constraint_data);
+            }
+        }
+
+        let mut mask = Vec::with_capacity(n_columns);
+        for i in 0..n_columns {
+            mask.push((0, i));
+        }
+
+        BoundaryAir {
+            trace_length,
+            n_columns,
+            constraints,
+            mask,
+        }
+    }
+}
+
+impl<F: PrimeField> Air<F> for BoundaryAir<F> {
+    fn trace_length(&self) -> usize {
+        self.trace_length
+    }
+
+    fn num_columns(&self) -> usize {
+        self.n_columns
+    }
+
+    fn num_random_coefficients(&self) -> usize {
+        self.constraints.len()
+    }
+
+    fn get_mask(&self) -> &[(usize, usize)] {
+        &self.mask
+    }
+
+    fn constraints_eval(
+        &self,
+        neighbors: &[F],
+        _periodic_columns: &[F],
+        random_coefficients: &[F],
+        point: &F,
+        _shifts: &[F],
+        _precomp_domains: &[F],
+    ) -> F {
```

Comment:
How do we compose this with other types of constraints?

Reply:
I haven't reviewed the Stone code where boundary constraints are composed with other types of constraints, so I can't answer this yet.